JBPM-6402: Stunner - Added support for profiles & implemented Ruleflow profile for BPMN

romartin opened the pull request:
Hey @hasys 

See https://issues.jboss.org/browse/JBPM-6402

This commit also is part of the task [JBPM-7605](https://issues.jboss.org/browse/JBPM-7605), as introduces support for generic "profiles" and the ability to switch between them. 

Also provides the concrete BPMN `RuleFlow` profile expected, and integrates the profile selection by using the profile specified on the workbench preferences.

Thanks!

romartin commented:
Jenkins execute full downstream build

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/menu/dev/impl/ChangeProfileDevCommand.java:
```diff
@@ -0,0 +1,68 @@
+import org.uberfire.client.views.pfly.modal.Bs3Modal;
+
+@Dependent
+public class ChangeProfileDevCommand extends AbstractMenuDevCommand {
```
Notice @hasys  this just a button for the "development" showcase, in order to test changing profiles from the designer, but this is not being included on the workbench distributions.

hasys commented:
Hi @romartin, sorry it is in a conflict now :(

romartin commented:
Hey @hasys , just rebased :+1:  Thanks!

romartin commented:
Jenkins execute full downstream build

romartin commented:
ping @hasys

hasys commented:
Hi @romartin, 

I fetched changes built and start showcase-project, but I can't see how to switch profiles 
![image](https://user-images.githubusercontent.com/1477262/50591054-dd732800-0e8d-11e9-9e07-e3027c184027.png)

romartin commented:
Hey @hasys 

No problem. If you recall we say to use the profile specified on the workbench preferences, so no need to use the showcase-project, you can use the complete business-central distribution and just go to the workbench preferences section, where you will see a "profiles" section, so you can change it there. Hope it helps! :+1: 

Thanks!

hasys commented:
ohh, sorry, @romartin,

build was already removed :( will trigger build again.

Jenkins execute full downstream build

jomarko commented:
@hasys please ping me when downs. build finish, I would like to check DMN part.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-backend-common/src/test/java/org/kie/workbench/common/stunner/core/backend/profile/BackendProfileManagerTest.java:
```diff
@@ -0,0 +1,94 @@
+    private Instance<Profile> profileInstances;
+
+    @Before
+    @SuppressWarnings("unchecked")
```
`SuppressWarnings` is redundant here.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-backend-common/src/test/java/org/kie/workbench/common/stunner/core/backend/profile/BackendProfileManagerTest.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Before
+    @SuppressWarnings("unchecked")
+    public void setup() throws Exception {
```
Method doesn't throw an exception.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/api/ClientProfileManagerTest.java:
```diff
@@ -0,0 +1,92 @@
+    private ManagedInstance<Profile> profileInstances;
+
+    @Before
+    @SuppressWarnings("unchecked")
```
Redundant `SuppressWarnings` here.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/api/ClientProfileManagerTest.java:
```diff
@@ -0,0 +1,92 @@
+    @Before
+    @SuppressWarnings("unchecked")
+    public void setup() throws Exception {
+        profileInstances = spy(new ManagedInstanceStub<Profile>(profile1, profile2));
```
Explicit type argument Profile can be replaced with <>.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/api/ProfileManager.java:
```diff
@@ -0,0 +1,53 @@
+    Collection<Profile> getAllProfiles();
+
+    /**
+     * Returns a given profiel by its identifier
```
Typo, should be:

_Returns a given **profile** by its identifier_

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/api/ProfileManager.java:
```diff
@@ -0,0 +1,53 @@
+    Collection<Profile> getProfiles(String definitionSetId);
+
+    /**
+     * Returns a given profiel by its identifier and its Definition Set domain.
```
Typo, should be:

_Returns a given **profile** by its identifier and its Definition Set domain._

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/profile/Profile.java:
```diff
@@ -0,0 +1,24 @@
+
+public interface Profile {
+
+    public String getProfileId();
```
Unnecessary `public` modifier.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/profile/Profile.java:
```diff
@@ -0,0 +1,24 @@
+
+    public String getProfileId();
+
+    public String getName();
```
Unnecessary `public` modifier.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/test/java/org/kie/workbench/common/stunner/core/lookup/domain/CommonDomainTargetNodeLookupTest.java:
```diff
@@ -0,0 +1,106 @@
+    private DomainLookupsCache cache;
+
+    private CommonDomainLookups tested;
+    private TestingGraphMockHandler graphTestHandler;
```
Field can be converted to a local variable.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/test/java/org/kie/workbench/common/stunner/core/lookup/domain/CommonDomainTargetNodeLookupTest.java:
```diff
@@ -0,0 +1,106 @@
+    private TestingGraphInstanceBuilder.TestGraph1 graph;
+
+    @Before
+    @SuppressWarnings("unchecked")
```
Redundant suppression.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/test/java/org/kie/workbench/common/stunner/core/lookup/domain/CommonDomainTargetNodeLookupTest.java:
```diff
@@ -0,0 +1,106 @@
+        CanConnect connectionRule = (CanConnect) ((List<Rule>) graphTestHandler.ruleSet.getRules()).get(0);
+        when(cache.getConnectionRules()).thenReturn(Collections.singletonList(connectionRule));
+        when(cache.getDefinitions(contains("label1"))).thenReturn(Arrays.asList(TestingGraphInstanceBuilder.DEF1_ID).stream().collect(Collectors.toSet()));
+        when(cache.getDefinitions(contains("label2"))).thenReturn(Arrays.asList(TestingGraphInstanceBuilder.DEF2_ID).stream().collect(Collectors.toSet()));
```
Can be simplified to:

`Stream.of(TestingGraphInstanceBuilder.DEF1_ID).collect(Collectors.toSet())`

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/test/java/org/kie/workbench/common/stunner/core/lookup/domain/CommonDomainTargetNodeLookupTest.java:
```diff
@@ -0,0 +1,106 @@
+        when(cache.getConnectionRules()).thenReturn(Collections.singletonList(connectionRule));
+        when(cache.getDefinitions(contains("label1"))).thenReturn(Arrays.asList(TestingGraphInstanceBuilder.DEF1_ID).stream().collect(Collectors.toSet()));
+        when(cache.getDefinitions(contains("label2"))).thenReturn(Arrays.asList(TestingGraphInstanceBuilder.DEF2_ID).stream().collect(Collectors.toSet()));
+        when(definitionsRegistry.getLabels(eq(TestingGraphInstanceBuilder.DEF1_ID))).thenReturn(TestingGraphInstanceBuilder.DEF1_LABELS);
```
Can be simplified to:

`Stream.of(TestingGraphInstanceBuilder.DEF2_ID).collect(Collectors.toSet())`

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/test/java/org/kie/workbench/common/stunner/core/profile/DomainProfileManagerTest.java:
```diff
@@ -0,0 +1,131 @@
+    private Object definitionSet;
+
+    @Mock
+    private Annotation domainQualifier;
```
This mock is never used.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/test/java/org/kie/workbench/common/stunner/core/profile/DomainProfileManagerTest.java:
```diff
@@ -0,0 +1,131 @@
+        when(adapterManager.forDefinitionSet()).thenReturn(definitionSetAdapter);
+        when(definitionSetAdapter.getDefinitions(eq(definitionSet)))
+                .thenReturn(Arrays.asList(DEF1, DEF2).stream().collect(Collectors.toSet()));
+        when(profileManager.getProfile(eq(DEF_SET_ID), eq(DEFAULT_PROFILE.getProfileId())))
```
Can be simplified to:

`Stream.of(DEF1, DEF2).collect(Collectors.toSet())`

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/preferences/StunnerProfilePreferencesLoaderTest.java:
```diff
@@ -0,0 +1,128 @@
+
+    @Before
+    @SuppressWarnings("unchecked")
+    public void setup() throws Exception {
```
Method doesn't produce an exception.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/preferences/StunnerProjectPreferencesRegistryLoaderTest.java:
```diff
@@ -0,0 +1,127 @@
+
+    @Before
+    @SuppressWarnings("unchecked")
+    public void setUp() throws Exception {
```
This method doesn't produce an exception.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/profile/ProfileSelector.java:
```diff
@@ -0,0 +1,95 @@
+        return profileManager.getProfile(id);
+    }
+}
+;
```
Redundant symbol.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/views/Selector.java:
```diff
@@ -0,0 +1,32 @@
+
+public interface Selector<T> {
+
+    public Selector<T> addItem(T item);
```
public modifier is redundant for interfaces.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/views/SelectorDelegateTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
```
Redundant suppression.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/views/SelectorDelegateTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
```
Redundant suppression.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/views/SelectorDelegateTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
```
Redundant suppression.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/views/SelectorDelegateTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
```
Redundant suppression.

romartin commented:
Hey @hasys 

Thanks for the review! :+1: 

Applied changes for all your comments, added the  `Reusable Subprocess` into the `RuleFlow` profile and also rebased with mater. Ready for review again!

Thanks!

romartin commented:
Jenkins execute full downstream build
